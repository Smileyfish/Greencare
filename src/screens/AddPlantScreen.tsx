import React, { useState, useEffect } from 'react';
import { View, Text, TextInput, TouchableOpacity, StyleSheet, Image, Alert } from 'react-native';
import * as ImagePicker from 'expo-image-picker';
import { addPlant } from '../services/database';
import { useNavigation } from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';

const AddPlantScreen: React.FC = () => {
    const [name, setName] = useState('');
    const [location, setLocation] = useState('');
    const [wateringInterval, setWateringInterval] = useState('');
    const [imageUri, setImageUri] = useState<string | null>(null);
    const [darkMode, setDarkMode] = useState(false);

    const navigation = useNavigation();

    useEffect(() => {
        const loadSettings = async () => {
            const dark = await AsyncStorage.getItem('darkMode');
            if (dark !== null) setDarkMode(dark === 'true');
        };
        loadSettings();
    }, []);

    const pickImage = async () => {
        const permissionResult = await ImagePicker.requestMediaLibraryPermissionsAsync();
        if (!permissionResult.granted) {
            Alert.alert('Berechtigung benötigt', 'Bitte erlauben Sie den Zugriff auf die Galerie.');
            return;
        }

        const result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.Images,
            quality: 0.5,
        });

        if (!result.canceled) {
            setImageUri(result.assets[0].uri);
        }
    };

    const takePhoto = async () => {
        const permissionResult = await ImagePicker.requestCameraPermissionsAsync();
        if (!permissionResult.granted) {
            Alert.alert('Berechtigung benötigt', 'Bitte erlauben Sie den Zugriff auf die Kamera.');
            return;
        }

        const result = await ImagePicker.launchCameraAsync({ quality: 0.5 });
        if (!result.canceled) {
            setImageUri(result.assets[0].uri);
        }
    };

    const savePlant = async () => {
        if (!name || !wateringInterval) {
            Alert.alert('Fehler', 'Name und Gießintervall sind Pflichtfelder.');
            return;
        }

        const interval = parseInt(wateringInterval, 10);
        if (isNaN(interval) || interval <= 0) {
            Alert.alert('Fehler', 'Gießintervall muss eine positive Zahl sein.');
            return;
        }

        try {
            await addPlant(name, location, interval, new Date().toISOString(), imageUri || '');
            navigation.goBack();
        } catch (error) {
            console.error('Failed to save plant:', error);
            Alert.alert('Fehler', 'Pflanze konnte nicht gespeichert werden. Bitte versuchen Sie es erneut.');
        }
    };

    return (
        <View style={[styles.container, darkMode && styles.darkContainer]}>
            <Text style={[styles.label, darkMode && styles.darkText]}>Name</Text>
            <TextInput
                style={[styles.input, darkMode && styles.darkInput]}
                value={name}
                onChangeText={setName}
                placeholder="Name"
                placeholderTextColor={darkMode ? '#aaa' : '#999'}
            />

            <Text style={[styles.label, darkMode && styles.darkText]}>Ort</Text>
            <TextInput
                style={[styles.input, darkMode && styles.darkInput]}
                value={location}
                onChangeText={setLocation}
                placeholder="Ort"
                placeholderTextColor={darkMode ? '#aaa' : '#999'}
            />

            <Text style={[styles.label, darkMode && styles.darkText]}>Gießintervall (Tage)</Text>
            <TextInput
                style={[styles.input, darkMode && styles.darkInput]}
                value={wateringInterval}
                onChangeText={setWateringInterval}
                keyboardType="numeric"
                placeholder="Tage"
                placeholderTextColor={darkMode ? '#aaa' : '#999'}
            />

            <View style={styles.buttonRow}>
                <TouchableOpacity style={styles.photoButton} onPress={takePhoto}>
                    <Text style={styles.photoButtonText}>Foto aufnehmen</Text>
                </TouchableOpacity>
                <TouchableOpacity style={styles.photoButton} onPress={pickImage}>
                    <Text style={styles.photoButtonText}>Galerie</Text>
                </TouchableOpacity>
            </View>

            {imageUri && <Image source={{ uri: imageUri }} style={styles.previewImage} />}

            <TouchableOpacity style={[styles.saveButton, darkMode && styles.darkSaveButton]} onPress={savePlant}>
                <Text style={styles.saveButtonText}>Pflanze speichern</Text>
            </TouchableOpacity>
        </View>
    );
};

const styles = StyleSheet.create({
    container: { flex: 1, padding: 16, backgroundColor: '#fff' },
    darkContainer: { backgroundColor: '#121212' },
    label: { fontSize: 16, fontWeight: 'bold', marginTop: 12, color: '#000' },
    darkText: { color: '#fff' },
    input: {
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 6,
        padding: 8,
        marginTop: 4,
        color: '#000',
        backgroundColor: '#fff',
    },
    darkInput: {
        borderColor: '#555',
        color: '#fff',
        backgroundColor: '#1E1E1E',
    },
    buttonRow: { flexDirection: 'row', justifyContent: 'space-between', marginTop: 16 },
    photoButton: {
        backgroundColor: '#4caf50',
        padding: 10,
        borderRadius: 6,
        flex: 1,
        marginHorizontal: 4,
    },
    photoButtonText: { color: 'white', textAlign: 'center' },
    previewImage: { width: 100, height: 100, marginTop: 16, borderRadius: 8 },
    saveButton: {
        backgroundColor: 'green',
        padding: 14,
        borderRadius: 8,
        marginTop: 24,
    },
    darkSaveButton: { backgroundColor: '#388E3C' },
    saveButtonText: { color: 'white', textAlign: 'center', fontSize: 16, fontWeight: 'bold' },
});

export default AddPlantScreen;
