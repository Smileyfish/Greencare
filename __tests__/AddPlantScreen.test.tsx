import React from "react";
import { render } from "@testing-library/react-native";
import AddPlantScreen from "../src/screens/AddPlantScreen";
import { NavigationContainer } from "@react-navigation/native";

// Mock AsyncStorage & Navigation
jest.mock("@react-native-async-storage/async-storage", () => ({
    getItem: jest.fn(() => Promise.resolve(null)),
    setItem: jest.fn(),
}));

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native");
    return {
        ...actualNav,
        useNavigation: () => ({ goBack: jest.fn(), navigate: jest.fn() }),
    };
});

describe("Snapshot: AddPlantScreen", () => {
    it("rendert korrekt", () => {
        const tree = render(
            <NavigationContainer>
                <AddPlantScreen />
            </NavigationContainer>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
