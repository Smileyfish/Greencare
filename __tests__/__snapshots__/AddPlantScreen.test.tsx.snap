// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: AddPlantScreen rendert korrekt 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
        "flex": 1,
        "padding": 16,
      },
      false,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginTop": 12,
        },
        false,
      ]
    }
  >
    Name
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="Name"
    placeholderTextColor="#999"
    style={
      [
        {
          "backgroundColor": "#fff",
          "borderColor": "#ccc",
          "borderRadius": 6,
          "borderWidth": 1,
          "color": "#000",
          "marginTop": 4,
          "padding": 8,
        },
        false,
      ]
    }
    value=""
  />
  <Text
    style={
      [
        {
          "color": "#000",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginTop": 12,
        },
        false,
      ]
    }
  >
    Ort
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="Ort"
    placeholderTextColor="#999"
    style={
      [
        {
          "backgroundColor": "#fff",
          "borderColor": "#ccc",
          "borderRadius": 6,
          "borderWidth": 1,
          "color": "#000",
          "marginTop": 4,
          "padding": 8,
        },
        false,
      ]
    }
    value=""
  />
  <Text
    style={
      [
        {
          "color": "#000",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginTop": 12,
        },
        false,
      ]
    }
  >
    Gie√üintervall (Tage)
  </Text>
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Tage"
    placeholderTextColor="#999"
    style={
      [
        {
          "backgroundColor": "#fff",
          "borderColor": "#ccc",
          "borderRadius": 6,
          "borderWidth": 1,
          "color": "#000",
          "marginTop": 4,
          "padding": 8,
        },
        false,
      ]
    }
    value=""
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 16,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#4caf50",
          "borderRadius": 6,
          "flex": 1,
          "marginHorizontal": 4,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Foto aufnehmen
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#4caf50",
          "borderRadius": 6,
          "flex": 1,
          "marginHorizontal": 4,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Galerie
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "green",
        "borderRadius": 8,
        "marginTop": 24,
        "opacity": 1,
        "padding": 14,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Pflanze speichern
    </Text>
  </View>
</View>
`;
